### PR TITLE
[dy2static] Support for static graph training with @declarative decorator

### PR DESCRIPTION
**Pull request by Aurelius84**
### 1. 背景
框架目前通过`@declarative`对动态图`forward`函数添加装饰，可以将网络转为静态图执行。但需要对`train`函数代码进行部分改写，放到`fluid.program_guard`语法下进行训练。

在动转静训练时，为了尽可能避免代码的修改，提升`@declarative`的易用性，将`@declarative`下的所有layers涉及的op转换为静态图模式执行。

image

### 2. 基本原理
 - 复用动态图的Trace执行流程，将被转为静态图的所有op，放到单独的子Program里
 - 解析子Program中Ops，组网并放到`run_program_op`进行trace，进行前向计算，并添加反向

如上图所示，`sub_layers`中的所有的ops计算都会放进`run_program_op`里。在执行sub_layers时，通过Trace `run_program_op`动态执行所有子Program前向ops。

在反向计算时，借助静态图的`fluid.gradients`接口，在子Program中添加反向grad_op。当执行`loss.backward()`时，会调用`run_program_op`的GradKernel，依次调用子Program中的`grad_op Kernel`计算grad。

**Example:**
```python
class Linear(fluid.dygraph.Layer):
    def __init__(self, input_dim=10, output_dim=5):
        super(Linear, self).__init__()
        self.fc = fluid.dygraph.Linear(input_dim, output_dim, act='relu')

    @declarative  # 添加装饰器
    def forward(self, x):
        pre = self.fc(x)
        loss = fluid.layers.mean(pre)
        return pre, loss

# 训练代码不需要做任何改动
def train(self, to_static=False):
    batch_num = 10
    with fluid.dygraph.guard(fluid.CPUPlace()):
        net = Linear()
        adam = fluid.optimizer.AdamOptimizer(learning_rate=0.1, parameter_list=net.parameters())
        
        for batch_id in range(batch_num):
            input = fluid.dygraph.to_variable(batch_data)
            pred, avg_loss = net(input)

            avg_loss.backward()
            adam.minimize(avg_loss)
            dygraph_net.clear_gradients()

    # 保存训练好的模型参数
    fluid.save_dygraph(net.state_dict(), "./test_dy2stat_save_load")

    infer_net = Linear()
    # 加载训练好的参数，进行预测
    model_dict, _ = fluid.load_dygraph("./test_dy2stat_save_load")
    infer_net.set_dict(model_dict)
    pred, _ = infer_net(x)
```

#### 新增
  - 针对旧`declarative`动转静训练，需手动改写`train`函数的问题，支持不需改写任何训练代码，只需在`forward`函数添加`declarative`即可转为静态图执行。
  - 支持在模型入口和局部`forward`函数添加`declarative`，转为静态子图执行。
  - 支持模型`train/eval`模式切换

#### 优化
 -  `run_program_op`的反向kernel计算后，将`TensorCopy`改为`ShareDataWith`，减少数据复制开销。
  - 针对python2下`convert_call`无法获取带装饰器的function原始code问题，进行修复，通过`__wrapped__`属性解析。
  - 针对输入参数强依赖于`to_variable(x)`语句来判断是否构建`feed layers`的问题，移除了对`to_variable`的依赖，可以通过`VarBase/Variable/Numpy.ndarry`进行解析创建。
  - 替换单测文件中`dygraph_to_static_func`为`declarative`，增加对`declarative`新逻辑的测试，逐步移除旧`dygraph_to_static_func`接口，包括MNIST、seResnet、ptb_lm动态图模型单测。
  - 添加动转静后`fluid.save_dygraph`和`fluid.load_dygraph`模型参数保存和加载后，预测结果的正确性。

#### 移除
   - 移除了`Program_translator`内的部分旧接口
      - 移除了旧的`set_optimizer`接口，支持训练代码不要用户修改，减少用户使用成本
      - 移除了`save_inference_model`接口，统一模型训练参数`save/load`与动态图使用方式一致，即可以直接使用`fluid.save_dygraph`和`fluid.load_dygraph`保存网络参数。
   - 移除了`convert_call`中对Python2和Python3的不同分支逻辑判断，精简了递归转化逻辑
    
TODO:
 - 暂不支持动静转换邻接layers传递LoDTensorArray类型的vars；
 - 不支持被装饰函数内定义带参数layer，会显示提示用户不允许（需将此部分layer放到`__init__`函数）

### CODE REVIEW

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py [hunk 1]
```diff
@@ -103,63 +103,63 @@ def dyfunc(x):
...
+                #     source_code = inspect.getsource(func)
+                #     if any(decorator in source_code
+                #            for decorator in DECORATOR_NAMES):
+                #         converted_call = None
+                #     else:
+                #         converted_call = to_static_func(func)
+                #         func_self = getattr(func, '__self__', None)
+                # else:
+                converted_call = to_static_func(func)
+                func_self = getattr(func, '__self__', None)
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py:107]: I will remove these code  in next commit.

File: python/paddle/fluid/dygraph/base.py [hunk 2]
```diff
@@ -61,6 +61,25 @@ def program_desc_tracing_guard(enable):
 _functional_dygraph_context_manager = None
 
 
+@signature_safe_contextmanager
+def check_param_type_guard(parameters):
```
Comment (zhhsplendid) [inline python/paddle/fluid/dygraph/base.py]: From code, I think it is like a `param_guard`, you copy the old parameters and put new_var into parameters. So what't the `check` and `type`? Is it proper to call the function `param_guard`?

File: python/paddle/fluid/dygraph/dygraph_to_static/ast_transformer.py [hunk 3]
```diff
@@ -146,6 +137,9 @@ def convert_to_static(dyfunc):
     Converts dygraph function into static function.
     """
     # Get AST from dygraph function
+    # Note: In Python2, it will raise OSError when inspect function
+    # with decorator directly and dyfunc.__wrapped__ holds the actual function.
+    dyfunc = getattr(dyfunc, '__wrapped__', dyfunc)
```
Comment (zhhsplendid) [inline python/paddle/fluid/dygraph/dygraph_to_static/ast_transformer.py:142]: Good finding!

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py (hunk 1 above)
Comment (zhhsplendid) [reply python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py:107]: Please ask Yamei for the if six.PY3 here. I remembered her did some changes for error cases for difference between Python2 and python3

File: python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            elif isinstance(value, core.VarBase):
+                var = value
+                var.name = self.inputs[i].desc.name()
+            else:
+                continue
+            input_vars.append(var)
+        # Create VarBase to receive output data.
+        out_vars = []
+        for var in self.outputs:
+            if not isinstance(var, framework.Variable): continue
```
Comment (zhhsplendid) [inline python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py]: Put continue at the next line

File: python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py [hunk 5]
```diff
@@ -391,25 +359,19 @@ def func(x):
...
-                else:
-                    outputs = tuple(outputs)
-        return outputs
+        function_spec = FunctionSpec(dygraph_func, args, kwargs)
+        _, partial_program_layer = self._program_cache[function_spec]
+
+        if args and isinstance(args[0], layers.Layer):
+            args = args[1:]
+
+        return partial_program_layer(args)
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py:374]: if user change`train` or `eval` mode when added @declarative, how to change partial_program_layer‘s mode?

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_save_inference_model.py [hunk 6]
```diff
@@ -44,7 +44,8 @@ def forward(self, x):
 
 
 class TestDyToStaticSaveInferenceModel(unittest.TestCase):
-    def test_save_inference_model(self):
+    # TODO(Aurelius84): disable temporarily, need new save_inference interface
```
Comment (zhhsplendid) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_save_inference_model.py:47]: Why? Is this due to the program is wrapped as dygraph layer so we cannot save inference model directly?

Comment (zhhsplendid) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_save_inference_model.py:47]: If so we should design new save methods. We should investigate whether save_dygraph is enough in our current case.

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py (hunk 1 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py:107]: We discussed offline. She splits into two case because `inpect.getsource(func)` raise a `OSError` when `func` is decorated in PY2. In this case, we should use `inspect.getsource(getattr(dyfunc, '__wrapped__'))` instead of `dyfunc` directly.

File: python/paddle/fluid/dygraph/base.py (hunk 2 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/base.py]: Thanks, `param_guard ` is a better name. `check_param_type_guard ` is used to switch parameters type: `ParamBase` --> `Parameter`.

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_save_inference_model.py (hunk 6 above)
Comment (Aurelius84) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_save_inference_model.py:47]: `save_inference_model` requires `executor`, but there is no `executor` in dygraph mode and it is inside `run_program_op`(C++).

The way of saving/ loading parameters of model is consistent with dygraph: `fluid.save_dygraph` and `fluid.load_dygraph`.

Also, we need a better way to implement `save_inference_model`.

File: python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py (hunk 5 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py:374]: Thanks, I I'm adding this feature...

File: python/paddle/fluid/dygraph/dygraph_to_static/ast_transformer.py (hunk 3 above)
Comment (liym27) [reply python/paddle/fluid/dygraph/dygraph_to_static/ast_transformer.py:142]: Great! And code about PY2 in `convert_func` can be removed

File: python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py (hunk 4 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py]: Thanks, fixed it.

File: python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py (hunk 5 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/dygraph_to_static/program_translator.py:374]: Done~~

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_mnist.py [hunk 7]
```diff
@@ -182,18 +182,18 @@ def train(self, to_static=False):
...
                             .format(epoch, batch_id,
                                     avg_loss.numpy(),
                                     acc.numpy(), time() - start))
                         start = time()
-                        if batch_id == 100:
-                            mnist.eval()
-                            prediction, acc, avg_loss = mnist(img, label)
-                            loss_data.append(avg_loss.numpy()[0])
-                            break
+                    if batch_id == 50:
```
Comment (liym27) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_mnist.py:192]: Not important. Better explain why it's changed to 50.

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_list.py [hunk 8]
```diff
@@ -138,11 +129,9 @@ def test_list_pop_in_if(x):
         a.append(x + 1)
         a.append(fluid.layers.fill_constant(shape=[2], value=2, dtype="int64"))
     item1 = a.pop(1)
-    a.pop()
```
Comment (liym27) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_list.py:141]: Better keep code `a.pop()`, because `a` is an empty list after `a.pop()` and it's a test case. Maybe add it next PR.

**Review by liym27 — DISMISSED** [review]
LGTM

**Review by chenwhql — DISMISSED** [review]
LGTM

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_mnist.py (hunk 7 above)
Comment (Aurelius84) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_mnist.py:192]: Before this PR, `test_mnist` only contains `train_static` but no comparison with the results of the dynamic graph.  I add `train_dygraph` and `assertEqual` in this unittest and reduce batch_id to 50 to make it run faster.
Another reason is I found the unittest will raise  random failure only `PR-CI-Py35` but it's ok on `PR-CI-Coverage` which is same platform `xly.bce.baidu`. I will check it for detail.

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_list.py (hunk 8 above)
Comment (Aurelius84) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_list.py:141]: Thanks, I will re-add it in next PR.

File: python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py (hunk 1 above)
Comment (zhhsplendid) [reply python/paddle/fluid/dygraph/dygraph_to_static/convert_call_func.py:107]: Great

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_save_inference_model.py (hunk 6 above)
Comment (zhhsplendid) [reply python/paddle/fluid/tests/unittests/dygraph_to_static/test_save_inference_model.py:47]: Sure. I will approve this PR, but we should discuss whether we can delete `save_inference_model` API

**Review by zhhsplendid — DISMISSED** [review]
LGTM

**Review by Xreki — DISMISSED** [review]
LGTM for the modification of framework.py. PR描述写的好详细，👍

File: python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+__all__ = ['partial_program_from', 'PartialProgramLayer']
+
+
+class PartialProgramLayer(layers.Layer):
+    """
+    PartialProgramLayer wraps all the ops from layers decorated by `@declarative`
+    and execute them as a static subgraph.
+
+    .. note::
+        **1. It should not be called directly and is used to train dygraph by static mode.
```
Comment (lanxianghit) [inline python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py:29]: 内部接口，请确认目前的写法是否会被暴露到官网上

**Review by lanxianghit — DISMISSED** [review]
PR描述写的非常好！

File: python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py (hunk 9 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/dygraph_to_static/partial_program.py:29]: Thanks，已经提commit修复了，移除了`__all__`避免内部接口暴露到外部用户。

**Review by liym27 — APPROVED** [review]
LGTM

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Review by Xreki — APPROVED** [review]

**Review by zhangting2020 — APPROVED** [review]
LGTM

**Review by Boyan-Liu — APPROVED** [review]

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]
LGTM
